### PR TITLE
feat: add azure blob filesystem + resourcesreader

### PR DESCRIPTION
**Pull request by EmanuelCampos**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [ ] Improvements to accept BaseFileSystemReader

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py [hunk 1]
```diff
@@ -64,102 +71,167 @@ class AzStorageBlobReader(BasePydanticReader):
...
+            download_file_path = os.path.join(temp_dir, sanitized_file_name)
+            logger.info(f"Start download of {self.blob}")
+            start_time = time.time()
+            with open(file=download_file_path, mode="wb") as download_file:
+                stream.readinto(download_file)
+            blob_meta[sanitized_file_name] = blob_client.get_blob_properties()
+            end_time = time.time()
+            logger.info(
+                f"{self.blob} downloaded in {end_time - start_time} seconds."
+            )
```
Comment (sourabhdesai) [inline llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py:102]: nit - maybe this should be a debug log?

File: llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py [hunk 2]
```diff
@@ -64,102 +71,167 @@ class AzStorageBlobReader(BasePydanticReader):
...
+        else:
+            logger.info("Listing blobs")
+            blobs_list = container_client.list_blobs(
+                self.name_starts_with, self.include
+            )
+            for obj in blobs_list:
```
Comment (sourabhdesai) [inline llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py:108]: Hmm wondering if we would we be able to de-duplicate the code in this method here by doing something like

```
if self.blob:
    blobs_list = [self.blob]
else:
    blobs_list = container_client.list_blobs(...)

for obj in blobs_list:
    ...
```

?

**Review by sourabhdesai — APPROVED** [review]

File: llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py
Comment (sourabhdesai) [inline llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py:1]: we should _really_ implement the async methods for this 😅 but again not a blocker

Comment (sourabhdesai) [reply llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py:1]: _(if the azstorage client lib exposes an async interface)_

**Review by sourabhdesai — APPROVED** [review]

**Comment by Cyronius** [comment]
this commit breaks passing in a single blob.  the optional parm 'blob' is a string, not a blob object, so when it loops through blobs_list, it immediately fails with "AttributeError: 'str' object has no attribute 'name'"

**Comment by SimZack** [comment]
Same issue as Cyronius :(

**Pull request merged**